File: services/token-revocation.js
/**
 * Token Revocation Service
 *
 * Database-backed JWT token revocation for HIPAA-compliant session management.
 * Stores SHA-256 hashes of revoked tokens (never the raw tokens).
 */

import { createHash } from 'crypto';
import { db } from '../db/client.js';
import { sql } from 'drizzle-orm';

function hashToken(token) {
  return createHash('sha256').update(token).digest('hex');
}

export const tokenRevocationService = {
  /**
   * Revoke a specific JWT token.
   * Stores the hash with a 7-day expiry (matching JWT max lifetime).
   */
  async revokeToken(token, revokedBy, reason = '') {
    const tokenHash = hashToken(token);
    await db.execute(sql`
      INSERT INTO revoked_tokens (token_hash, revoked_by, reason, expires_at)
      VALUES (${tokenHash}, ${revokedBy}, ${reason}, NOW() + INTERVAL '7 days')
      ON CONFLICT (token_hash) DO NOTHING
    `);
  },

  /**
   * Revoke all tokens for a given admin (bulk revocation).
   * Stores a marker row keyed by admin ID.
   */
  async revokeAllForAdmin(adminId, revokedBy, reason = '') {
    const markerHash = createHash('sha256').update(`revoke_all:${adminId}`).digest('hex');
    await db.execute(sql`
      INSERT INTO revoked_tokens (token_hash, revoked_by, reason, expires_at)
      VALUES (${markerHash}, ${revokedBy}, ${reason || `revoke_all for admin ${adminId}`}, NOW() + INTERVAL '7 days')
      ON CONFLICT (token_hash) DO UPDATE SET
        revoked_at = NOW(),
        revoked_by = EXCLUDED.revoked_by,
        reason = EXCLUDED.reason,
        expires_at = EXCLUDED.expires_at
    `);
  },

  /**
   * Check if a specific token has been revoked. Fast indexed lookup.
   */
  async isTokenRevoked(token) {
    const tokenHash = hashToken(token);
    const result = await db.execute(sql`
      SELECT 1 FROM revoked_tokens
      WHERE token_hash = ${tokenHash} AND expires_at > NOW()
    `);
    return result.rows.length > 0;
  },

  /**
   * Check if all tokens for an admin have been bulk-revoked.
   */
  async isAdminRevoked(adminId) {
    const markerHash = createHash('sha256').update(`revoke_all:${adminId}`).digest('hex');
    const result = await db.execute(sql`
      SELECT 1 FROM revoked_tokens
      WHERE token_hash = ${markerHash} AND expires_at > NOW()
    `);
    return result.rows.length > 0;
  },

  /**
   * Remove expired revocation entries. Returns count deleted.
   */
  async cleanupExpired() {
    const result = await db.execute(sql`
      WITH d AS (DELETE FROM revoked_tokens WHERE expires_at < NOW() RETURNING 1)
      SELECT count(*) AS c FROM d
    `);
    return parseInt(result.rows[0]?.c || '0', 10);
  },
};

File: pipecat/services/token_revocation.py
"""Token revocation service.

Provides database-backed JWT token revocation for HIPAA-compliant session management.
Stores SHA-256 hashes of revoked tokens (never the raw tokens).
"""

from __future__ import annotations

import hashlib

from loguru import logger

from db.client import query_one, query_many, execute


def _hash_token(token: str) -> str:
    """SHA-256 hash a JWT token for storage."""
    return hashlib.sha256(token.encode()).hexdigest()


async def revoke_token(token: str, revoked_by: str, reason: str = "") -> None:
    """Revoke a specific JWT token.

    Stores the hash with a 7-day expiry (matching JWT max lifetime).
    """
    token_hash = _hash_token(token)
    await execute(
        """INSERT INTO revoked_tokens (token_hash, revoked_by, reason, expires_at)
           VALUES ($1, $2, $3, NOW() + INTERVAL '7 days')
           ON CONFLICT (token_hash) DO NOTHING""",
        token_hash,
        revoked_by,
        reason,
    )
    logger.info(
        "Revoked token by={who} reason_chars={n}",
        who=str(revoked_by)[:8],
        n=len(reason or ""),
    )


async def revoke_all_for_admin(admin_id: str, revoked_by: str, reason: str = "") -> int:
    """Revoke all active tokens for a given admin.

    Since JWTs are stateless and we can't enumerate them, this stores a
    marker row. The auth middleware must also check admin-level revocation.
    Returns count of newly inserted revocation rows (always 1 for the marker).
    """
    marker_hash = hashlib.sha256(f"revoke_all:{admin_id}".encode()).hexdigest()
    await execute(
        """INSERT INTO revoked_tokens (token_hash, revoked_by, reason, expires_at)
           VALUES ($1, $2, $3, NOW() + INTERVAL '7 days')
           ON CONFLICT (token_hash) DO UPDATE SET
             revoked_at = NOW(),
             revoked_by = EXCLUDED.revoked_by,
             reason = EXCLUDED.reason,
             expires_at = EXCLUDED.expires_at""",
        marker_hash,
        revoked_by,
        reason or f"revoke_all for admin {admin_id}",
    )
    logger.info("Revoked all tokens for admin={aid}", aid=str(admin_id)[:8])
    return 1


async def is_token_revoked(token: str) -> bool:
    """Check if a specific token has been revoked. Fast indexed lookup."""
    token_hash = _hash_token(token)
    row = await query_one(
        "SELECT 1 FROM revoked_tokens WHERE token_hash = $1 AND expires_at > NOW()",
        token_hash,
    )
    return row is not None


async def is_admin_revoked(admin_id: str) -> bool:
    """Check if all tokens for an admin have been revoked (bulk revocation)."""
    marker_hash = hashlib.sha256(f"revoke_all:{admin_id}".encode()).hexdigest()
    row = await query_one(
        "SELECT 1 FROM revoked_tokens WHERE token_hash = $1 AND expires_at > NOW()",
        marker_hash,
    )
    return row is not None


async def cleanup_expired() -> int:
    """Remove expired revocation entries. Returns count deleted."""
    result = await query_one(
        "WITH d AS (DELETE FROM revoked_tokens WHERE expires_at < NOW() RETURNING 1) SELECT count(*) AS c FROM d"
    )
    count = result["c"] if result else 0
    if count > 0:
        logger.info("Cleaned up {n} expired token revocations", n=count)
    return count
